### PR TITLE
[REVIEW] Improve Peak Finding function performance

### PR DESCRIPTION
**Pull request by mnicely**
# Info

This ticket is to optimize function in the peak finding module.

To do so, CUDA kernel will be added for 1D and 2D arrays, while 3D+ arrays will continue to use CuPy functions.

We see a significant improvement as the order increases.

### branch-0.16 (1D)
```bash
--------------- benchmark 'Argrelextrema': 12 tests ---------------
Name (time in us)                                    Mean          
-------------------------------------------------------------------
test_argrelextrema_gpu[1-wrap-1--1-32768]        276.5078 (1.41)   
test_argrelextrema_gpu[1-clip-1--1-32768]        334.0082 (1.70)   
test_argrelextrema_gpu[1-wrap-2--1-32768]        395.9963 (2.02)   
test_argrelextrema_gpu[1-clip-2--1-32768]        507.2813 (2.58)   
test_argrelextrema_gpu[1-wrap-25--1-32768]     3,129.2780 (15.93)  
test_argrelextrema_gpu[1-clip-25--1-32768]     4,520.6792 (23.01)  
-------------------------------------------------------------------

--------------- benchmark 'Argrelmin': 12 tests ---------------
Name (time in us)                                Mean          
---------------------------------------------------------------
test_argrelmin_gpu[1-wrap-1--1-32768]        280.1709 (1.40)   
test_argrelmin_gpu[1-clip-1--1-32768]        334.6817 (1.67)   
test_argrelmin_gpu[1-wrap-2--1-32768]        400.1712 (2.00)   
test_argrelmin_gpu[1-clip-2--1-32768]        510.7902 (2.55)   
test_argrelmin_gpu[1-wrap-25--1-32768]     3,147.4148 (15.73)  
test_argrelmin_gpu[1-clip-25--1-32768]     4,694.4378 (23.46)  
---------------------------------------------------------------

------------- benchmark 'TestArgrelmax': 12 tests -------------
Name (time in us)                                Mean          
---------------------------------------------------------------
test_argrelmax_gpu[1-wrap-1--1-32768]        281.2009 (1.45)   
test_argrelmax_gpu[1-clip-1--1-32768]        344.2331 (1.77)   
test_argrelmax_gpu[1-wrap-2--1-32768]        405.2330 (2.08)   
test_argrelmax_gpu[1-clip-2--1-32768]        511.3321 (2.63)   
test_argrelmax_gpu[1-wrap-25--1-32768]     3,134.2374 (16.12)  
test_argrelmax_gpu[1-clip-25--1-32768]     4,509.1346 (23.19)  
---------------------------------------------------------------
```

### This PR (1D)
```bash
--------------- benchmark 'Argrelextrema': 12 tests ---------------
Name (time in us)                                    Mean          
-------------------------------------------------------------------
test_argrelextrema_gpu[1-clip-25--1-32768]       143.4626 (1.0)    
test_argrelextrema_gpu[1-clip-1--1-32768]        143.6572 (1.00)   
test_argrelextrema_gpu[1-wrap-2--1-32768]        144.2907 (1.01)   
test_argrelextrema_gpu[1-wrap-25--1-32768]       145.0372 (1.01)   
test_argrelextrema_gpu[1-wrap-1--1-32768]        145.4891 (1.01)   
test_argrelextrema_gpu[1-clip-2--1-32768]        145.8152 (1.02)   
-------------------------------------------------------------------

--------------- benchmark 'Argrelmin': 12 tests ---------------
Name (time in us)                                Mean          
---------------------------------------------------------------
test_argrelmin_gpu[1-wrap-1--1-32768]        144.2479 (1.0)    
test_argrelmin_gpu[1-clip-1--1-32768]        145.8922 (1.01)   
test_argrelmin_gpu[1-clip-25--1-32768]       146.3002 (1.01)   
test_argrelmin_gpu[1-wrap-2--1-32768]        146.4851 (1.02)   
test_argrelmin_gpu[1-clip-2--1-32768]        146.5781 (1.02)   
test_argrelmin_gpu[1-wrap-25--1-32768]       147.9346 (1.03)   
---------------------------------------------------------------

------------- benchmark 'TestArgrelmax': 12 tests -------------
Name (time in us)                                Mean          
---------------------------------------------------------------
test_argrelmax_gpu[1-clip-25--1-32768]       144.6847 (1.0)    
test_argrelmax_gpu[1-clip-1--1-32768]        145.3602 (1.00)   
test_argrelmax_gpu[1-clip-2--1-32768]        145.9188 (1.01)   
test_argrelmax_gpu[1-wrap-1--1-32768]        145.9430 (1.01)   
test_argrelmax_gpu[1-wrap-25--1-32768]       146.3621 (1.01)   
test_argrelmax_gpu[1-wrap-2--1-32768]        146.7437 (1.01)   
---------------------------------------------------------------
```

### branch-0.16 (2D)
```bash

-------------- benchmark 'Argrelextrema': 6 tests --------------
Name (time in us)                                 Mean          
----------------------------------------------------------------
test_argrelextrema_gpu[2-wrap-1-0-256]        319.4117 (1.0)    
test_argrelextrema_gpu[2-clip-1-0-256]        374.3033 (1.17)   
test_argrelextrema_gpu[2-wrap-2-0-256]        487.3124 (1.53)   
test_argrelextrema_gpu[2-clip-2-0-256]        588.0762 (1.84)   
test_argrelextrema_gpu[2-wrap-25-0-256]     3,996.5381 (12.51)  
test_argrelextrema_gpu[2-clip-25-0-256]     5,286.1737 (16.55)  
----------------------------------------------------------------

-------------- benchmark 'Argrelmin': 6 tests --------------
Name (time in us)                             Mean          
------------------------------------------------------------
test_argrelmin_gpu[2-wrap-1-0-256]        322.3578 (1.0)    
test_argrelmin_gpu[2-clip-1-0-256]        373.7319 (1.16)   
test_argrelmin_gpu[2-wrap-2-0-256]        489.2302 (1.52)   
test_argrelmin_gpu[2-clip-2-0-256]        580.2760 (1.80)   
test_argrelmin_gpu[2-wrap-25-0-256]     4,022.7967 (12.48)  
test_argrelmin_gpu[2-clip-25-0-256]     5,323.6967 (16.51)  
------------------------------------------------------------

------------ benchmark 'TestArgrelmax': 6 tests ------------
Name (time in us)                             Mean          
------------------------------------------------------------
test_argrelmax_gpu[2-wrap-1-0-256]        321.1711 (1.0)    
test_argrelmax_gpu[2-clip-1-0-256]        388.7706 (1.21)   
test_argrelmax_gpu[2-wrap-2-0-256]        481.5804 (1.50)   
test_argrelmax_gpu[2-clip-2-0-256]        608.2789 (1.89)   
test_argrelmax_gpu[2-wrap-25-0-256]     3,975.4080 (12.38)  
test_argrelmax_gpu[2-clip-25-0-256]     5,692.5087 (17.72)  
------------------------------------------------------------
```

### This PR (2D)
```bash
------------- benchmark 'Argrelextrema': 6 tests -------------
Name (time in us)                               Mean          
--------------------------------------------------------------
test_argrelextrema_gpu[2-wrap-2-0-256]      118.7306 (1.0)    
test_argrelextrema_gpu[2-clip-1-0-256]      119.2056 (1.00)   
test_argrelextrema_gpu[2-clip-2-0-256]      120.2280 (1.01)   
test_argrelextrema_gpu[2-clip-25-0-256]     120.7775 (1.02)   
test_argrelextrema_gpu[2-wrap-25-0-256]     120.8319 (1.02)   
test_argrelextrema_gpu[2-wrap-1-0-256]      121.1888 (1.02)   
--------------------------------------------------------------

------------- benchmark 'Argrelmin': 6 tests -------------
Name (time in us)                           Mean          
----------------------------------------------------------
test_argrelmin_gpu[2-wrap-1-0-256]      120.0341 (1.0)    
test_argrelmin_gpu[2-wrap-25-0-256]     120.1067 (1.00)   
test_argrelmin_gpu[2-clip-1-0-256]      120.7193 (1.01)   
test_argrelmin_gpu[2-wrap-2-0-256]      121.3296 (1.01)   
test_argrelmin_gpu[2-clip-2-0-256]      121.9923 (1.02)   
test_argrelmin_gpu[2-clip-25-0-256]     122.3397 (1.02)   
----------------------------------------------------------

----------- benchmark 'TestArgrelmax': 6 tests -----------
Name (time in us)                           Mean          
----------------------------------------------------------
test_argrelmax_gpu[2-wrap-1-0-256]      119.6535 (1.0)    
test_argrelmax_gpu[2-clip-2-0-256]      119.7168 (1.00)   
test_argrelmax_gpu[2-wrap-2-0-256]      120.6676 (1.01)   
test_argrelmax_gpu[2-wrap-25-0-256]     120.8693 (1.01)   
test_argrelmax_gpu[2-clip-25-0-256]     121.9161 (1.02)   
test_argrelmax_gpu[2-clip-1-0-256]      124.3351 (1.04)   
----------------------------------------------------------

```

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by awthomp** [comment]
@mnicely -- Is this PR ready to be reviewed? I keep seeing commits pushed and want to make sure I don't jump the gun!

**Comment by mnicely** [comment]
> @mnicely -- Is this PR ready to be reviewed? I keep seeing commits pushed and want to make sure I don't jump the gun!

Sorry about that, I'm having CI issues with the CUDA 10.X. SMH. Trying to fix now!

**Comment by mnicely** [comment]
> @mnicely -- Is this PR ready to be reviewed? I keep seeing commits pushed and want to make sure I don't jump the gun!


Give it a go!

File: CHANGELOG.md [hunk 1]
```diff
@@ -21,6 +21,8 @@
 - PR #222 - Improved performance for various window functions
 - PR #235 - Improve Wavelets functions performance
 - PR #242 - Add PyTorch disclaimer to notebook
+- PR #243 - Improve Peak Finding function performance
+- PR #242 - Add PyTorch disclaimer to notebook
```
Comment (awthomp) [inline CHANGELOG.md:25]: This is a super nit, but can you switch the order of these PR lists? I.e. put 242 before 243?

Comment (mnicely) [reply CHANGELOG.md:25]: Yup!

**Review by awthomp — CHANGES_REQUESTED** [review]
One minor nit on PR ordering, but looks great overall. Awesome work!

**Review by awthomp — APPROVED** [review]

**Pull request merged**